src/slurm_plugin/cluster_event_publisher.py

```diff
@@ -0,0 +1,371 @@
+            "After node maintenance, node currently in replacement",
+            event_type="static-node-in-replacement",
+            timestamp=timestamp,
+            event_supplier=self._node_in_replacement_supplier(
+                (node for node in unhealthy_static_nodes if node.name in nodes_in_replacement)
+            ),
+        )
+
+    @log_exception(logger, "publish_nodes_failing_health_check_events", catch_exception=Exception, raise_on_error=False)
+    def publish_nodes_failing_health_check_events(
+        self, health_check_type: any, node_names_failing_health_check: Iterable[str]
+    ):
+        def detail_supplier():
+            node_names = list(node_names_failing_health_check)
+            if node_names_failing_health_check:
+                yield {
+                    "detail": {
+                        "health-check-type": str(health_check_type),
+                        "count": len(node_names_failing_health_check),
+                        "nodes": [{"name": node_name} for node_name in self._limit_list(node_names)],
+                    }
+                }
+
+        timestamp = ClusterEventPublisher.timestamp()
+
+        self.publish_event(
+            "WARNING",
+            f"Number of nodes failing health check {health_check_type}",
+            "nodes-failing-health-check-count",
+            timestamp=timestamp,
+            event_supplier=detail_supplier(),
+        )
+
+    @log_exception(logger, "publish_unhealthy_node_events", catch_exception=Exception, raise_on_error=False)
+    def publish_unhealthy_node_events(self, unhealthy_nodes: List[SlurmNode]):
+        def detail_supplier():
+            invalid_backing_instance_nodes = [
+                node for node in unhealthy_nodes if not node.is_backing_instance_valid(log_warn_if_unhealthy=False)
+            ]
+            if invalid_backing_instance_nodes:
+                yield {
+                    "detail": {
+                        "count": len(invalid_backing_instance_nodes),
+                        "nodes": self._generate_node_name_list(invalid_backing_instance_nodes),
+                    }
+                }
+
+        timestamp = ClusterEventPublisher.timestamp()
+
+        self.publish_event(
+            "WARNING",
+            "Number of nodes without a valid backing instance",
+            "invalid-backing-instance-count",
+            timestamp=timestamp,
+            event_supplier=detail_supplier(),
+        )
+
+    @log_exception(logger, "publish_bootstrap_failure_events", catch_exception=Exception, raise_on_error=False)
+    def publish_bootstrap_failure_events(self, bootstrap_failure_nodes: List[SlurmNode]):
+        def dynamic_node_filter(node: SlurmNode) -> bool:
+            return isinstance(node, DynamicNode) and node.is_resume_failed() and node.is_nodeaddr_set()
+
+        def static_node_filter(node: SlurmNode) -> bool:
+            return isinstance(node, StaticNode) and node._is_replacement_timeout
```

Comment:
Since these are the logic in is_bootstrap_failure, what about we instead define a function inside class slurm node `is_bootstrap_timeout(self)` that have different implementation for static node and dynamic node, and the function can be reused in `is_bootstrap_failure` and here?

Reply:
Yeah, I can do that. I was just trying to minimize changes to working code.

Reply:
Yes, this is much cleaner - thank you!